segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeepertier2/StorageLedger.java

```diff
@@ -0,0 +1,183 @@
+     * @param offset offset from which writes start.
+     * @return The metadata of the ledger.
+     */
+    public CompletableFuture<LedgerData> getLedgerDataForWriteAt(long offset) {
+        CompletableFuture<LedgerData> retVal = new CompletableFuture<>();
+        if (offset != length) {
+            retVal.completeExceptionally(new BadOffsetException(this.getName(), length, offset));
+        }
+        LedgerData ledgerData = this.getLastLedgerData();
+        if (ledgerData != null && !ledgerData.getLh().isClosed()) {
+            retVal.complete(ledgerData);
+        } else {
+            /** If there is no ledger, create a new one. */
+            return manager.createLedgerAt(this.name, (int) offset).thenApply(data -> {
+                this.dataMap.put((int) offset, data);
+                return data;
+        });
+        }
+        return retVal;
+    }
+
+    /**
+     * Add a new BK ledger and metadata at a given offset.
+     * @param offset The starting offset represented by the ledger.
+     * @param ledgerData metadata of the ledger.
+     */
+    public synchronized void addToList(int offset, LedgerData ledgerData) {
+
+        /** If we are replacing an existing ledger, adjust the length */
+
+        LedgerData older = this.dataMap.put(offset, ledgerData);
+        if (older != null) {
+            this.length -= older.getLh().getLength();
+        }
+        this.length += ledgerData.getLh().getLength();
+    }
+
+    /**
+     * Increase length of the StorageLedger as a side effect of the write operation.
+     * This is just a cache operation. The length is not persisted.
+     *
+     * @param size size of data written.
+     */
+    public synchronized void increaseLengthBy(int size) {
+        this.length += size;
+    }
+
+
+    public synchronized CompletableFuture<Void> deleteAllLedgers() {
+        return CompletableFuture.allOf(
+                this.dataMap.entrySet().stream().map(entry -> manager.deleteLedger(entry.getValue().getLh())).toArray(CompletableFuture[]::new));
+    }
+
+    public LedgerData getLastLedgerData() {
+        if (this.dataMap.isEmpty()) {
+            return null;
+        } else {
+            return this.dataMap.entrySet().stream().max((entry1, entry2) -> entry1.getKey() - entry2.getKey()).get().getValue();
+        }
+    }
+
+    public void setSealed() {
+        sealed = true;
+    }
+
+    /**
+     * Creates a list of curator transaction for merging source StorageLedger in to this.
+     * @param source Name of the source ledger.
+     * @return list of curator operations.
+     */
+    public List<CuratorOp> addLedgerDataFrom(StorageLedger source) {
+        List<CuratorOp> retVal = source.dataMap.entrySet().stream().map(entry -> {
+            int newKey = entry.getKey() + this.length;
+            this.dataMap.put(newKey, entry.getValue());
+            return manager.createAddOp(this.name, newKey, entry.getValue());
+        }).collect(Collectors.toList());
```

Comment:
can `createAddOp` fail? If so, you may have corrupted `dataMap` since you only partially updated it.

Reply:
The actual execution of the op created by `createAllOp` may fail. This is taken care of by the caller where it drops the cache.